test/validation/api/ipsec/ipsec.c

```diff
@@ -891,18 +977,33 @@ static int ipsec_send_out_one(const ipsec_test_part *part,
 
 				pkto[i] = odp_packet_from_event(ev);
 				CU_ASSERT_FATAL(pkto[i] != ODP_PACKET_INVALID);
+
+				if (part->out[i].sa_expiry != IPSEC_TEST_EXPIRY_NONE)
+					ipsec_status_event_get(sa, part->out[i].sa_expiry);
+
 				i++;
 				continue;
 			}
 
 			ev = recv_event(suite_context.queue, 0);
 			if (ODP_EVENT_INVALID != ev) {
+				odp_event_type_t ev_type;
+
 				CU_ASSERT(odp_event_is_valid(ev) == 1);
+				ev_type = odp_event_types(ev, &subtype);
+
+				if ((ODP_EVENT_IPSEC_STATUS == ev_type) &&
+				    part->out[i].sa_expiry != IPSEC_TEST_EXPIRY_NONE) {
+					ipsec_status_event_handle(ev, sa, part->out[i].sa_expiry);
+					continue;
+				}
+
 				CU_ASSERT_EQUAL(ODP_EVENT_PACKET,
-						odp_event_types(ev, &subtype));
+						ev_type);
 				CU_ASSERT_EQUAL(ODP_EVENT_PACKET_IPSEC,
 						subtype);
-				CU_ASSERT(part->out[i].status.error.all);
+				if (part->out[i].sa_expiry == IPSEC_TEST_EXPIRY_NONE)
```

Comment:
I do not see the reason for this change. We should not get an event in the SA queue unless there was an error (and thus some error flag must be set) or the event is a status event, but status event have already been handled. So this change looks unnecessary to me. If it is not, maybe it deserves a clarifying comment?

Reply:
When expiry is being tested (i.e, sa_expiry != IPSEC_TEST_EXPIRY_NONE), eventually ipsec_check_sa_expiry() gets invoked from ipsec_check_out(), the caller of this function (ipsec_send_out_one). In ipsec_check_sa_expiry(), the error bits are getting checked as expected. I can add comment as below to clarify this:

```
+
+                               /* In the case of SA hard expiry tests, hard expiry error bits are
+                                * expected to be set. The exact error bits expected to be set based
+                                * on sa_expiry is checked eventually in ipsec_check_sa_expiry()
+                                * from the caller of this function.
+                                */
```

Let me know if this works.

Reply:
I am not sure the comment really helps to understand why the assertion should be skipped when the expiry test is ongoing. The real reason seems to be that the expected error (part->out[i].status.error.all) is not set to match what is really expected (and in some cases the expected result is that an expiry status may or may not be set).

This is actually part of the bigger issue that the added expiry check does not quite follow the earlier pattern that test cases set the test packet and the expected output in the test_part and then let ipsec.c just do the processing and checking without knowing avout the test case. This patch blurs that a bit by having expiry test specific logic inside ipsec.c. OTOH, part of that is a bit unavoidable since SA expiry reporting (especially with soft limits) is not exact.